### ISSUE TITLE
ODataBatch: Allow individual requests and change sets in batch requests.

### ISSUE THREAD

**Issue by krombipils** [object/Simple.OData.Client#139]
Maybe I missed something, but in the current implementation all requests added to an ODataBatch are part of a single change set. Regarding to the OData specification a batch request 'is made up of a series of individual requests and change sets'.
I would really appreciate the possibility to mix individual requests and change sets.

**Comment by object** [comment]
I've added this one to the backlog. Perhaps you can give me an example of how you benefit from having multiple changesets in a batch (ideally sketching it with Simple.OData.Client API syntax).

**Comment by vladascibulskis** [comment]
My scenario for this question would be:
1. There are multiple records that must be processed
2. Each record has to be processed executing multiple OData API calls
3. Record processing must either succeed fully or rollback on failure (changeset)
4. Each individual record can succeed or fail to process independently (no changesets).

Sample code:

    IEnumerable<Flight> flight1 = null;
    IEnumerable<Flight> flight2 = null;
    
    var batch = new ODataBatch(client);
    batch += c => c
        .For<Flight>()
        .Key(flight1.Key)
        .Action("ReserveSeat")
        .ExecuteAsync()
        .InChangeSet("1");
    batch += c => c
        .For<PaymentProcessor>()
        .Action("PayForFlight")
        .Set(new { Flight = flight1.Key, CreditCard = card })
        .ExecuteAsync()
        .InChangeSet("1");
    batch += c => c
        .For<Flight>()
        .Key(flight2.Key)
        .Action("ReserveSeat")
        .ExecuteAsync()
        .InChangeSet("2");
    batch += c => c
        .For<PaymentProcessor>()
        .Action("PayForFlight")
        .Set(new { Flight = flight2.Key, CreditCard = card })
        .ExecuteAsync()
        .InChangeSet("2");
    await batch.ExecuteAsync();

**Comment by germrfy** [comment]
I am also interested in this functionality.

I have a use case similar to @indyjason79 where I wish to create multiple (000's) entities at once.
In the event one of those fails they will all fail and I do not want to have to retry each one individually.

**Comment by phatcher** [comment]
Does anyone know if the Microsoft OData library supports this, we can add a syntax that would push the data but if none of the server side libraries know how to process it it won't mean anything.

**Comment by vladascibulskis** [comment]
Yes it is supported by MS OData server side library, and with some additional tweaking (ex. https://damienbod.com/2014/08/14/web-api-odata-v4-batching-part-10/) change-set can be executed in single DB transaction, while non change-set requests can be executed in separate DB transactions. 
DB transaction scoping is one of the main reasons for me, why I would like to have this functionality.

**Comment by phatcher** [comment]
@vladascibulskis Thanks for that - it makes much easier to get to a working prototype.

**Comment by josh-hoiland** [comment]
Another reason this would be good to add is that some services like Dynamics 365 don't support multiple requests per changeset. 
```
{
  "error": {
    "code": "BadRequest_NotSupported",
    "message": "Multiple requests within the same change set are not supported by Dynamics 365 Business Central OData web services.  CorrelationId:  2ab595b5-0262-42af-be11-adc96c108f60."
  }
}
```

**Comment by calebshannon** [comment]
+1 for change set allocation.

Our very specific use case is recreating the behaviour of the Dyanmics 365 F&O Excel Add In. Each request is in a separate changeset, providing "attempt-all" behaviour.

With Simple.OData I am only able to get "all or nothing" behaviour when all requests are put in a single changeset.

Primarily wanted to show some support/activity on this issue, but also wondering if anyone has a workaround.
